File: ios/GhostChat/Core/Crypto/IdentityKeyService.swift
import CryptoKit
import Foundation

/// Persistent P-256 identity keypair that survives app restarts. Generated once on first
/// call and retained in the Keychain. The public key is what peers see as "who I am"
/// for safety-number comparison and pending-room HOST/GUEST determination.
final class IdentityKeyService: @unchecked Sendable {

    enum Keys {
        static let privateRaw = "identity.private.raw"
    }

    enum Error: Swift.Error, Equatable {
        case decodingFailed
    }

    private let keychain: KeychainServicing
    private let lock = NSLock()
    private var cached: P256.KeyAgreement.PrivateKey?

    init(keychain: KeychainServicing) {
        self.keychain = keychain
    }

    /// Returns the identity private key, generating it on first call.
    func getOrCreateIdentity() throws -> P256.KeyAgreement.PrivateKey {
        lock.lock(); defer { lock.unlock() }
        if let c = cached { return c }
        if let data = try keychain.get(Keys.privateRaw) {
            guard let key = try? P256.KeyAgreement.PrivateKey(rawRepresentation: data) else {
                throw Error.decodingFailed
            }
            cached = key
            return key
        }
        let fresh = P256.KeyAgreement.PrivateKey()
        try keychain.set(fresh.rawRepresentation, for: Keys.privateRaw)
        cached = fresh
        return fresh
    }

    /// 65-byte x963 representation (includes 0x04 prefix).
    var publicKeyX963: Data {
        get throws { try getOrCreateIdentity().publicKey.x963Representation }
    }

    /// 64-byte raw coordinate (x + y), used in safety-number comparisons.
    var publicKeyRaw: Data {
        get throws {
            let x963 = try publicKeyX963
            return Data(x963.dropFirst())
        }
    }

    /// Irrevocably deletes the identity. Used by panic-wipe.
    func resetIdentity() throws {
        lock.lock(); defer { lock.unlock() }
        cached = nil
        try keychain.delete(Keys.privateRaw)
    }
}

File: android/app/src/main/java/com/kordar/ghostchat/core/crypto/IdentityKeyService.kt
package com.kordar.ghostchat.core.crypto

import com.kordar.ghostchat.core.crypto.CryptoUtils as Bc
import com.kordar.ghostchat.core.security.KeystoreServicing

/**
 * Persistent P-256 identity keypair. Generated once on first call, then retained in the
 * [KeystoreServicing]-backed secure store. Mirrors iOS `IdentityKeyService`: the public
 * key represents "who am I" for safety-number comparison and for the pending-room
 * HOST/GUEST determination.
 */
class IdentityKeyService(
    private val keystore: KeystoreServicing
) {

    object Keys {
        const val PRIVATE_RAW = "identity.private.raw"
    }

    @Volatile private var cached: Bc.ECKeyPair? = null

    /** Get or lazily create the identity keypair. Thread-safe. */
    @Synchronized
    fun getOrCreateIdentity(): Bc.ECKeyPair {
        cached?.let { return it }
        val stored = keystore.get(Keys.PRIVATE_RAW)
        val keypair = if (stored != null) {
            Bc.keyPairFromPrivateBytes(stored)
        } else {
            Bc.generateKeyPair().also { fresh ->
                keystore.set(Keys.PRIVATE_RAW, fresh.privateKeyBytes)
            }
        }
        cached = keypair
        return keypair
    }

    /** 65-byte uncompressed public (0x04 || X || Y). */
    val publicKeyX963: ByteArray
        get() = getOrCreateIdentity().publicKeyBytes

    /** 64-byte raw public (X || Y) for safety-number input. */
    val publicKeyRaw: ByteArray
        get() = getOrCreateIdentity().publicKeyRaw

    /** Irrevocably deletes the identity. Called during panic wipe. */
    @Synchronized
    fun resetIdentity() {
        cached = null
        keystore.delete(Keys.PRIVATE_RAW)
    }
}
